[BoundsSafety] Parse external bounds attributes in ObjC methods

hnrklssn opened the pull request:
This parses counted_by et. al. for parameters and return types in Objective-C method signatures. No semantic checking is done to ensure that protocol, interface and implementation are aligned.

rdar://145190177

hnrklssn committed:
[BoundsSafety] Parse external bounds attributes in ObjC methods

This parses counted_by et. al. for parameters and return types in
Objective-C method signatures. No semantic checking is done to ensure
that protocol, interface and implementation are aligned.

rdar://145190177

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please test llvm

apple-fcloutier commented on clang/test/BoundsSafety/AST/objc.m:
```diff
@@ -0,0 +1,154 @@
+// CHECK-NEXT:   |-ObjCInterface {{.*}} 'CountedByClass'
+// CHECK-NEXT:   |-ObjCMethodDecl {{.*}} - simpleParam:: 'void'
+// CHECK-NEXT:   | |-ImplicitParamDecl {{.*}} self 'CountedByClass *'
+// CHECK-NEXT:   | |-ImplicitParamDecl {{.*}} _cmd 'SEL':'SEL *'
```
We should probably consciously decide what to do with `SEL`.

hnrklssn replied:
As in, we should make it `__single`? I guess that would apply to `self` also.

apple-fcloutier replied:
`SEL` is _typically_ a C string. I don't know for sure what constraints the runtime puts on the contents of the string, but @mikeash should know.

`self` (like all other ObjC pointers) should always be `__single`, and I think that one is un-controversial. I seem to recall that ObjC pointers are not actually PointerTypes, though?

mikeash replied:
Currently, `SEL` is always a C string. We reserve the right to change that in the future (there are some advantages to be had by making some common selectors into small integers), but I'm skeptical that we'd be able to actually pull it off, there has to be a lot of code out there that casts to `char *` and expects it to work.

The only constraints on the contents are that they must be NUL terminated and unique, i.e. you're not allowed to have two distinct `SEL` values that both point to bytewise-equal strings. And they must remain valid for the lifetime of the process. The runtime ensures these things if you get your `SEL` by legal means.

hnrklssn replied:
Right. Is `SEL *` a `char *` or `char **`? Do we ever cast `_cmd`/`*_cmd` to `char *`, or is it ~always accessed through something like `NSStringFromSelector`? Would making it `__single` in the public interface but `__null_terminated` behind the scenes be feasible? I guess I'm wondering how much we expect user code to access the selector internals.

mikeash replied:
`SEL` is a funny `const char *`, so a `SEL *` would be a funny `const char **`. The runtime casts to `char *`. Nothing else is supposed to (the only legal means to get a selector's string is with `sel_getName` or something that wraps it), but I bet there's a decent amount of code out there that does. But it shouldn't.

Is there something even more restrictive than `__single`? Outside the runtime, code should never examine the pointed-to memory at all. Otherwise, `__single` seems fine.

apple-fcloutier replied:
We have `__sized_by(0)`. On one hand, it's impossible to dereference, which is good, but on the other hand, it's impossible to dereference, which is bad. For anyone currently doing it, what would you say that the fix is? Use `sel_getName`?

mikeash replied:
Yes, `sel_getName` is the right way to get the contents. Right now it’s a glorified cast internally, but it’s guaranteed to give you a readable C string somehow.

apple-fcloutier replied:
From my read, it's not a huge burden to support SEL as a `__sized_by(0)` pointer (or even as a pointer to an incomplete type); we can have a special diagnostic to direct people to use `sel_getName` if we fail a cast from `SEL` to `char *`.

(We don't have to do it in this PR, but if that sounds like the way to go, we should file a radar for it)

mikeash replied:
Sounds good. Breaking code that tries to directly cast a selector is fine by me.

hnrklssn replied:
not that it'll matter much for a non-dereferenceable pointer, but just to double check: `_cmd` is never null, right?

mikeash replied:
It’s unlikely, but possible. You’re allowed to get the IMP of a method and call it with whatever arguments you want.

hnrklssn replied:
Good news: there's already a warning for casting `SEL`: `-Wcast-of-sel-type`, and ObjCObjectPointerType already forbids pointer arithmetic, so no need to make `self` `__single`.

Bad news: 
it's kinda tricky to get a good user experience with `__sized_by(0)` for `SEL`, because `__sized_by` is not currently supported in typedefs. I can force it into the type hierarchy anyways, but since `__sized_by` is type sugar, it doesn't appear in the desugared type, but it's also hidden by the typedef in the sugared form. So `typedef SEL * __sized_by(0) SEL` gets printed as `'SEL':'SEL*'` under attribute only mode, and (even more confusingly), `'SEL':'SEL* __single'` under full `-fbounds-safety`.
We also don't have very good support for diagnosing unconditional traps from `__sized_by(0)` casts or derefs. So this is more likely to cause pain than anything else imo.

I tried getting around some of these problems by using `typedef SEL SEL[0]` instead, as a zero-sized array will decay to `__counted_by(0)` while still being printed in the desugared type, but there are many places assuming that the selector type is a pointer type. Similarly, making it not a typedef would require some changes as well.

Considering the current use of this will mainly be in protocols and interfaces in headers for Swift interop, rather than actually enforcing bounds safety in the ObjC class implementation, I think I'll punt annotating `_cmd` for now.

rapidsna replied:
Yes, we should track this issue separately. Also, fwiw, I don't want "attribute-only" mode to add any implicit annotation. It should only be able to parse explicit annotations. So, this maybe relevant for the experimental objc bounds-safety mode.

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test

rapidsna commented on clang/test/BoundsSafety/AST/objc.m:
```diff
@@ -0,0 +1,154 @@
+// CHECK-NEXT:   | |-ParmVarDecl {{.*}} len 'int'
+// CHECK-NOT: IsDeref
+// CHECK-NEXT:   | | `-DependerDeclsAttr
+// CHECK-NEXT:   | `-ParmVarDecl {{.*}} p 'int *{{(__single)?}} __counted_by(len)':'int *{{(__single)?}}'
```
We don't make thin pointers  `__single` in attribute-only modes, so the expected output should be different between the two modes.

hnrklssn committed:
[BoundsSafety] Update objc test case (NFC)

By separating the check lines containing implicit __single, we
explicitly check that attribute-only mode does not contain any implicit
__single attributes, while -fbounds-safety does.

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please test llvm

rapidsna reviewed (approved):
LGTM

hnrklssn commented:
@swift-ci please test linux

hnrklssn commented:
@swift-ci please test macos
